src/fsharp/PostInferenceChecks.fs

```diff
@@ -479,23 +479,102 @@ let CheckNoReraise cenv freesOpt (body:Expr) =
 /// Check if a function is a quotation splice operator
 let isSpliceOperator g v = valRefEq g v g.splice_expr_vref || valRefEq g v g.splice_raw_expr_vref 
 
-/// Check conditions associated with implementing multiple instantiations of a generic interface
-let CheckMultipleInterfaceInstantiations cenv interfaces m = 
-    let keyf ty = assert isAppTy cenv.g ty; (tcrefOfAppTy cenv.g ty).Stamp
-    let table = interfaces |> MultiMap.initBy keyf
-    let firstInterfaceWithMultipleGenericInstantiations = 
-        interfaces |> List.tryPick (fun typ1 -> 
-            table |> MultiMap.find (keyf typ1) |> List.tryPick (fun typ2 -> 
-                   if // same nominal type
-                       tyconRefEq cenv.g (tcrefOfAppTy cenv.g typ1) (tcrefOfAppTy cenv.g typ2) &&
-                       // different instantiations
-                       not (typeEquivAux EraseNone cenv.g typ1 typ2) 
-                    then Some (typ1,typ2)
-                    else None))
-    match firstInterfaceWithMultipleGenericInstantiations with 
+
+/// Examples:
+/// I<int> & I<int> => ExactlyEqual.
+/// I<int> & I<string> => NotEqual.
+/// I<int> & I<'T> => ``Equal except for TType_var or Measure or Alias``.
+/// with '[<Measure>] type kg': I< int<kg> > & I<int> => ``Equal except for TType_var or Measure or Alias``.
+/// with 'type MyInt = int': I<MyInt> & I<int> => ``Equal except for TType_var or Measure or Alias``.
+///
+/// The differences could also be nested, example:
+/// I<List<int*string>> vs I<List<int*'T>> => ``Equal except for TType_var or Measure or Alias``.
+type TTypeEquality =
+| ExactlyEqual
+| ``Equal except for TType_var or Measure or Alias``
+| NotEqual
+
+let compareTypesWithRegardToTypeVariablesAndMeasures g typ1 typ2 =
```

Comment:
this is probably way too complicated, I'll need to rething this some time later

Reply:
Yes, this looks a bit too complicated. Without these checks, the compiler already protects us against interfaces that may unify for type parameter substitutions; just the error message isn't as good as it says "Duplicate or redundant interface".

What I would do is go to the place where the compiler does the check for duplicates, in `MethodOverrides.fs`. If the duplicate error is about to thrown, determine if you should throw a different error based on the information that you have.

Reply:
It looks like this comment needs to be addressed... I don't want to add another complicated implementation of type equality logic to the compiler.  I'll take a look and see if I can make the adjustment utilising code already available.